ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs

```diff
@@ -267,30 +290,136 @@ data Ap :: (Type -> Type) -> LedgerStateKind -> Type -> Constraint -> Type where
-        tickThenApply cfg b l
+      withBlockReadSets b $ \lh ->
+        either (throwLedgerError db r) return $ runExcept $
+          tickThenApply cfg b lh
     Weaken ap' ->
       applyBlock cfg ap' db
   where
-    l :: l ValuesMK
-    l = error "UTxO HD applyBlock" (ledgerDbCurrent db :: l EmptyMK)
+    withBlockReadSets
+      ::HasDiskDb m l
+      => blk
+      -> (l ValuesMK -> m (l TrackingMK))
+      -> m (l TrackingMK)
+    withBlockReadSets b f = do
+      let ks = getKeySets b Nothing :: TableKeySets l
+      let aks = rewindTableKeySets (ledgerDbChangelog db) ks :: RewoundTableKeySets l
+      urs <- readDb aks
+      withHydratedLedgerState urs f
+
+    withHydratedLedgerState
+      :: UnforwardedReadSets l
+      -> (l ValuesMK -> a)
+      -> a
+    withHydratedLedgerState urs f =
+      case forwardTableKeySets (ledgerDbChangelog db) urs of
+        Nothing ->
+          -- We should explain here in which circumstances this might happen.
+          --
+          -- Here we're doing the rewind;read;forward sequence using the same
+          -- ledger state (and without flushing to disk), so the forward
+          -- operation __must__ succeed. However, what happens if the database
+          -- becomes corrupted. Is this realistic?
+          --
+          -- Also, in the future, when we introduce pipelining, the
+          -- rewind;read;forward sequence will not take place in 'applyBlock':
+          -- block keys reads will occur way before we invoke chain selection
+          -- for that block. So does it still make sense to have a __fallback__
+          -- rewind;read;forward sequence in 'applyBlock'?
+          --
+          error "TODO: handle this case appropriately."
+          -- TODO: right way to handle seems to be:
+          --
+          -- > re-issue rewind;read;fw
+          --
+        Just rs ->
+          f $ withLedgerTables (ledgerDbCurrent db)  rs
+
+{-------------------------------------------------------------------------------
+  HD Interface that I need (Could be moved to  Ouroboros.Consensus.Ledger.Basics )
+-------------------------------------------------------------------------------}
+
+
+data DbChangelog (l :: LedgerStateKind)
+  deriving (Eq, Generic, NoThunks)
+
+newtype RewoundTableKeySets l = RewoundTableKeySets (AnnTableKeySets l ()) -- KeySetSanityInfo l
+
+initialDbChangelog
+  :: WithOrigin SlotNo -> l EmptyMK -> DbChangelog l
+initialDbChangelog = undefined
+
+rewindTableKeySets
+  :: DbChangelog l -> TableKeySets l -> RewoundTableKeySets l
+rewindTableKeySets = undefined
+
+newtype UnforwardedReadSets l = UnforwardedReadSets (AnnTableReadSets l ())
+
+forwardTableKeySets
+  :: DbChangelog l -> UnforwardedReadSets l -> Maybe (TableReadSets l)
+forwardTableKeySets = undefined
+
+extendDbChangelog
+  :: WithOrigin SlotNo
+    -- TODO: use SeqNo and delegate WithOrigin SlotNo to SeqNo transformation to the instances of HasSeqNo class.
+  -> l DiffMK
+  -> Maybe (l SnapshotsMK)
```

Comment:
This arg could be:

```haskell
Maybe (LedgerTables l SnapshotsMK)
```

Otherwise `l` would represent the entire ledger state.